autoPyTorch/pipeline/components/setup/network_head/fully_connected.py

```diff
@@ -59,23 +59,38 @@ def get_hyperparameter_search_space(
     ) -> ConfigurationSpace:
         cs = ConfigurationSpace()
 
-        min_num_layers, max_num_layers = num_layers.value_range
-        num_layers_hp = get_hyperparameter(num_layers, UniformIntegerHyperparameter)
+        num_layers_is_constant = True if len(num_layers.value_range) == 1 else False
+        min_num_layers: int = num_layers.value_range[0]  # type: ignore
+        max_num_layers: int = num_layers.value_range[-1]  # type: ignore
 
+        num_layers_hp = get_hyperparameter(num_layers, UniformIntegerHyperparameter)
         activation_hp = get_hyperparameter(activation, CategoricalHyperparameter)
+        cs.add_hyperparameter(num_layers_hp)
 
-        cs.add_hyperparameters([num_layers_hp, activation_hp])
-        cs.add_condition(CS.GreaterThanCondition(activation_hp, num_layers_hp, 1))
+        if num_layers_is_constant:
+            # only add activation if we have more than 1 layer
+            if max_num_layers > 1:
+                cs.add_hyperparameter(activation_hp)
+        else:
+            cs.add_hyperparameter(activation_hp)
+            cs.add_condition(CS.GreaterThanCondition(activation_hp, num_layers_hp, 1))
 
-        for i in range(1, int(max_num_layers)):
-            num_units_search_space = HyperparameterSearchSpace(hyperparameter=f"units_layer_{i}",
-                                                               value_range=units_layer.value_range,
-                                                               default_value=units_layer.default_value,
-                                                               log=units_layer.log)
+        for i in range(1, max_num_layers + 1):
+            num_units_search_space = HyperparameterSearchSpace(
+                hyperparameter=f"units_layer_{i}",
+                value_range=units_layer.value_range,
+                default_value=units_layer.default_value,
+                log=units_layer.log,
+            )
             num_units_hp = get_hyperparameter(num_units_search_space, UniformIntegerHyperparameter)
-            cs.add_hyperparameter(num_units_hp)
 
-            if i >= int(min_num_layers):
-                cs.add_condition(CS.GreaterThanCondition(num_units_hp, num_layers_hp, i))
+            if i >= min_num_layers:
+                cs.add_hyperparameter(num_units_hp)
+                # In the case of a constant, the max and min number of layers are the same.
+                # So no condition is needed. If it is not a constant but a hyperparameter,
+                # then a condition has to be made so that it accounts for the value of the
+                # hyperparameter.
+                if not num_layers_is_constant:
+                    cs.add_condition(CS.GreaterThanCondition(num_units_hp, num_layers_hp, i))
```

Comment:
```suggestion
            cs.add_hyperparameter(num_units_hp)
            # (1) n_layers is const ==> n_units for each layer is always required (not conditional)
            # (2) n_layers is not const ==> n_units for deeper layers is required only when
            #     the corresponding layers are appended in the architecture. (conditional)
            if not num_layers_is_constant:
                cs.add_condition(CS.GreaterThanCondition(num_units_hp, num_layers_hp, i))
```

Reply:
why would you like to remove this `if i >= min_num_layers:` condition?

Reply:
Good point!
Thanks to your comment, I checked this point again.
So the last version before addressing my comment is [here](https://github.com/ravinkohli/Auto-PyTorch/blob/040d26214bdd907b8ea2db3603df19574b51d2a0/autoPyTorch/pipeline/components/setup/network_head/fully_connected.py#L78-L94).

The question is:
1. Do we set `num_units_hp` for all the layers or just from `min_num_layers` to `max_num_layers`?
2. If we do not set `num_units_hp` from `1` to `min_num_layers - 1`, why do we iterate from `1`?

I thought that we do not set this hyperparameter for lower layers, but if I think through more, it is a bit strange to not set this hyperparameter from the `1` to `min_num_layers - 1`.
Means shouldn't we have something like this?
```
        for i in range(1, max_num_layers + 1):
            num_units_search_space = HyperparameterSearchSpace(
                hyperparameter=f"units_layer_{i}",
                value_range=units_layer.value_range,
                default_value=units_layer.default_value,
                log=units_layer.log,
            )
            num_units_hp = get_hyperparameter(num_units_search_space, UniformIntegerHyperparameter)
            cs.add_hyperparameter(num_units_hp)

            if i >= min_num_layers and not num_layers_is_constant:
                # In the case of a constant, the max and min number of layers are the same.
                # So no condition is needed. If it is not a constant but a hyperparameter,
                # then a condition has to be made so that it accounts for the value of the
                # hyperparameter.
                cs.add_condition(CS.GreaterThanCondition(num_units_hp, num_layers_hp, i))
```

Reply:
we set num_units for all the layers, 
we iterate from 1 as we count from 1 for the layers so the first layer has num_units_1 rather than num_units_0.

Yeah I agree with the code you have presented. Thanks for pointing this out, I think this was not correct before, we should search for all the units and actually I think the current code will also fail if we have min_num_layers >1 as the `self.config` wont have units for the layers less than the min. I'll fix this.